### PR TITLE
define visibility class API

### PR DESCRIPTION
**Pull request by annavolp**
This pull request defines a new object for visibilities and its API.

### CODE REVIEW

File: xrayvision/visibility.py [hunk 1]
```diff
@@ -39,8 +152,8 @@ class Visibility:
         The x, y offset of phase center
 
     """
-    @apu.quantity_input(uv=1/apu.arcsec, offset=apu.arcsec, center=apu.arcsec, pixel_size=apu.arcsec)
-    def __init__(self, vis, *, u, v, offset=(0., 0.) * apu.arcsec, center=(0., 0.) * apu.arcsec):
+    @apu.quantity_input(u=1/apu.arcsec, v=1/apu.arcsec)
+    def __init__(self, visibilities, *, u, v, names, uncertainty=None, meta=None):
```
Comment (DanRyanIrish) [inline xrayvision/visibility.py:156]: I think this is what we need to make sure all the inputs are the correct types.  We should confirm this is correct though.

```suggestion
    @apu.quantity_input()
    def __init__(self,
                 visibilities: apu.Quantity,
                 *,
                 u: apu.Quantity[1/apu.arcsec],
                 v: apu.Quantity[1/apu.arcsec],
                 names: Iterable[str],
                 uncertainty: Any = None,
                 meta=None):
```

**Comment by DanRyanIrish** [comment]
@annavolp Can you push your latest work including the tests you wrote in Dublin?

File: xrayvision/visibility.py [hunk 2]
```diff
@@ -5,10 +5,409 @@
...
+    def __init__(
+        self,
+        visibilities: apu.Quantity,
+        u: apu.Quantity[1 / apu.arcsec],
+        v: apu.Quantity[1 / apu.arcsec],
+        names: Iterable[str],
+        meta: VisMetaABC,
+        dims: Iterable[str] = ("uv",),
+        uncertainty: Union[apu.Quantity, None],
+        coords: dict = {},
```
Comment (DanRyanIrish) [inline xrayvision/visibility.py:338]:
```suggestion
        coords: dict = {}
```

File: xrayvision/visibility.py [hunk 3]
```diff
@@ -5,10 +5,409 @@
...
+    def __init__(
+        self,
+        visibilities: apu.Quantity,
+        u: apu.Quantity[1 / apu.arcsec],
+        v: apu.Quantity[1 / apu.arcsec],
+        names: Iterable[str],
+        meta: Any = None,
+        dims: Iterable[str] = ("uv",),
+        uncertainty: Union[apu.Quantity, None],
+        coords: dict = {},
```
Comment (DanRyanIrish) [inline xrayvision/visibility.py:152]:
```suggestion
        coords: dict = {}
```

**Review by DanRyanIrish — CHANGES_REQUESTED** [review]

File: xrayvision/visibility.py [hunk 4]
```diff
@@ -5,10 +5,409 @@
...
+    @apu.quantity_input()
+    def __init__(
+        self,
+        visibilities: apu.Quantity,
+        u: apu.Quantity[1 / apu.arcsec],
+        v: apu.Quantity[1 / apu.arcsec],
+        names: Iterable[str],
+        meta: Any = None,
+        dims: Iterable[str] = ("uv",),
+        uncertainty: Union[apu.Quantity, None],
```
Comment (DanRyanIrish) [inline xrayvision/visibility.py:151]:
```suggestion
        uncertainty: Union[apu.Quantity, None] = None,
```

File: xrayvision/visibility.py [hunk 5]
```diff
@@ -5,10 +5,409 @@
...
+    @apu.quantity_input()
+    def __init__(
+        self,
+        visibilities: apu.Quantity,
+        u: apu.Quantity[1 / apu.arcsec],
+        v: apu.Quantity[1 / apu.arcsec],
+        names: Iterable[str],
+        meta: VisMetaABC,
+        dims: Iterable[str] = ("uv",),
+        uncertainty: Union[apu.Quantity, None],
```
Comment (DanRyanIrish) [inline xrayvision/visibility.py:337]:
```suggestion
        uncertainty: Union[apu.Quantity, None] = None,
```

**Review by DanRyanIrish — CHANGES_REQUESTED** [review]

**Comment by DanRyanIrish** [comment]
If we agree that #62 is ready to be merged now, it probably makes sense to combined that PR into this one first and then merge both Anna and my contributions together.  I think we're basically there.  Of course, small changes can still be made in the future if we want.

**Review by DanRyanIrish — APPROVED** [review]

**Comment by DanRyanIrish** [comment]
@samaloney This is ready to merge.

File: xrayvision/visibility.py [hunk 6]
```diff
@@ -5,10 +5,408 @@
...
+
 import astropy.units as apu
 import numpy as np
+import xarray
+from astropy.coordinates import SkyCoord
+from astropy.time import Time
+
+__all__ = ["Visibility", "Visibilities", "VisMeta", "VisibilitiesABC", "VisMetaABC"]
+
+_E_RANGE_KEY = "energy_range"
```
Comment (DanRyanIrish) [inline xrayvision/visibility.py:20]:
```suggestion
_E_RANGE_KEY = "spectral_range"
```

File: xrayvision/visibility.py [hunk 7]
```diff
@@ -5,10 +5,408 @@
...
+    def observer_coordinate(self) -> Union[SkyCoord, None]:
+        """
+        Location of the observer.
+        """
+
+    @property
+    @abc.abstractmethod
+    def energy_range(self) -> Union[Iterable[apu.Quantity], None]:
+        """
+        Energy range over which the visibilities are computed.
```
Comment (DanRyanIrish) [inline xrayvision/visibility.py:39]:
```suggestion
    def spectral_range(self) -> Union[Iterable[apu.Quantity], None]:
        """
        Spectral range over which the visibilities are computed.
```

File: xrayvision/visibility.py [hunk 8]
```diff
@@ -5,10 +5,408 @@
...
+            raise KeyError(f"Inputs must include a key, '{key}', that gives a {key_type}.")
+        if unit is not None and value is not None and not value.unit.is_equivalent(unit, equivalencies=equivalencies):
+            raise ValueError(f"'{key}' must have angular units.")
+
+    @property
+    def observer_coordinate(self):
+        return self.get(_OBS_COORD_KEY, None)
+
+    @property
+    def energy_range(self):
```
Comment (DanRyanIrish) [inline xrayvision/visibility.py:391]:
```suggestion
    def spectral_range(self):
```

**Review by DanRyanIrish — CHANGES_REQUESTED** [review]

**Comment by DanRyanIrish** [comment]
@samaloney Actually last minute change.  We should make "energy" "spectral".

**Review by samaloney — APPROVED** [review]

**Pull request merged**